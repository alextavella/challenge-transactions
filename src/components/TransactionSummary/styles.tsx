import styled from 'styled-components';
import { Container as MaterialContainer, Typography } from '@material-ui/core';

export const Container = styled(MaterialContainer).attrs({
  maxWidth: 'sm',
})`
  > div + div {
    margin-top: 24px;
  }
`;

export const TopicText = styled(Typography).attrs({
  variant: 'body2',
  component: 'p',
})``;

export const ValueText = styled(Typography).attrs({
  variant: 'subtitle2',
  component: 'strong',
  color: 'secondary',
})``;
