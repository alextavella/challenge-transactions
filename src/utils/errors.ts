import { ValidationError } from 'yup';

interface Error {
  [key: string]: string;
}

export function getValidationErrors(errors: ValidationError): any {
  return errors.inner.reduce((acc: Error, error: ValidationError) => {
    return { ...acc, [error.path]: error.message };
  }, {});
}
