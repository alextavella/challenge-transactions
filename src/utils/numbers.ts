export const formatCurrency = (value: number): string => {
  const formatted = new Intl.NumberFormat([], {
    style: 'currency',
    currency: 'BRL',
    minimumFractionDigits: 2,
  }).format(value);

  return formatted
    .replace('.', '#')
    .replace(/,/g, '.')
    .replace('#', ',')
    .replace('R$', 'R$ ');
};

export const unformatCurrency = (value: string): number => {
  return Number(
    value
      .replace('R$ ', '')
      .replace(',', '#')
      .replace(/\./g, '')
      .replace('#', '.'),
  );
};
