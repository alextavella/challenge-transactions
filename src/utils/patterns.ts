import { ValidationRules } from 'react-hook-form';

export interface MaskRule {
  rules: ValidationRules;
  transform: (text: string) => string;
}

export const CPF: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{3}).([0-9]{3}).([0-9]{3})-([0-9]{2})$/),
      message: 'CPF inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,3})(\d{0,3})(\d{0,3})(\d{0,2})/)!;

    const part1 = result[2] ? `${result[1]}.${result[2]}` : result[1];
    const part2 = result[3] ? `.${result[3]}` : '';
    const part3 = result[4] ? `-${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_NUMBER: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{4})\s([0-9]{4})\s([0-9]{4})\s([0-9]{4})$/),
      message: 'Número do cartão inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,4})(\d{0,4})(\d{0,4})(\d{0,4})/)!;

    const part1 = result[2] ? `${result[1]} ${result[2]}` : result[1];
    const part2 = result[3] ? ` ${result[3]}` : '';
    const part3 = result[4] ? ` ${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_EXPIRATION_DATE: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^(0[1-9]|1[012])\/?(20[2-9][0-9])$/),
      message: 'Data de expiração inválida',
    },
  },
  transform: (text: string = ''): string => {
    const result = text.replace(/\D/g, '').match(/(\d{0,2})(\d{0,4})/)!;

    return result[2] ? `${result[1]}/${result[2]}` : result[1];
  },
};

export const CREDIT_CARD_CVV: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/[0-9]{3}/),
      message: 'CVV inválido',
    },
  },
  transform: (text: string = ''): string => {
    return text.substr(0, 3);
  },
};

export const CURRENCY: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/([0-9]\.?)?[0-9]+,[0-9]{2}/),
      message: 'Valor inválido',
    },
  },
  transform: (text: string = ''): string => {
    let number = text.replace(/[^0-9]/g, '');

    if (number.length === 0) number = '0.00';
    else if (number.length === 1) number = '0.0' + number;
    else if (number.length === 2) number = '0.' + number;
    else
      number =
        number.substring(0, number.length - 2) +
        '.' +
        number.substring(number.length - 2, number.length);

    number = Number(number).toFixed(2);
    number = number.replace(/\./g, ',');

    const x = number.split(',');
    const x1 = x[0];
    const x2 = x.length > 1 ? ',' + x[1] : '';

    const formatAmountNoDecimals = (number: string) => {
      var rgx = /(\d+)(\d{3})/;
      while (rgx.test(number)) {
        number = number.replace(rgx, `$1.$2`);
      }
      return number;
    };

    return `R$ ` + formatAmountNoDecimals(x1) + x2;
  },
};
