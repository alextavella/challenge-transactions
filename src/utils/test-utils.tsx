import { render } from '@testing-library/react';
import React from 'react';
import MainProvider from '../providers';

const AllTheProviders: React.FC = ({ children }) => {
  return <MainProvider>{children}</MainProvider>;
};

const customRender = (ui: React.ReactElement, options?: any) =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
