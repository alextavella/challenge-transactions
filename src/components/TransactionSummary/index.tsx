import React from 'react';
import { formatCurrency } from '../../utils/numbers';
import { Container, TopicText, ValueText } from './styles';

interface TransactionSummaryProps {
  quantity: number;
  total: number;
}

const TransactionSummary: React.FC<TransactionSummaryProps> = ({
  quantity,
  total,
}) => {
  return (
    <Container>
      <div>
        <TopicText>Número de transações</TopicText>
        <ValueText data-testid="transaction_header_quantity">
          {quantity}
        </ValueText>
      </div>
      <div>
        <TopicText>Valor total</TopicText>
        <ValueText data-testid="transaction_header_total">
          {formatCurrency(total)}
        </ValueText>
      </div>
    </Container>
  );
};

export default TransactionSummary;
