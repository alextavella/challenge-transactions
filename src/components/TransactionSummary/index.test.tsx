import React from 'react';
import TransactionSummary from '.';
import { cleanup, render, screen } from '../../utils/test-utils';

describe('<TransactionSummary />', () => {
  afterEach(() => {
    cleanup();
  });

  it('should be able to render correctly', () => {
    // Arrange
    const { baseElement } = render(
      <TransactionSummary quantity={123} total={456} />,
    );

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to render the text values correctly', () => {
    // Arrange
    render(<TransactionSummary quantity={123} total={456} />);

    // Act
    const quantityText = screen.getByTestId(
      'transaction_header_quantity',
    ) as HTMLElement;

    const totalText = screen.getByTestId(
      'transaction_header_total',
    ) as HTMLElement;

    // Assert
    expect(quantityText.textContent).toBe('123');
    expect(totalText.textContent).toBe('R$ 456,00');
  });
});
