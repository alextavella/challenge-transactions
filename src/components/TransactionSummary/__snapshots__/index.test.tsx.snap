// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionSummary /> should be able to render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root sc-bdfBwQ eKkLOU MuiContainer-maxWidthSm"
    >
      <div>
        <p
          class="MuiTypography-root sc-gsTCUz MuiTypography-body2"
        >
          Número de transações
        </p>
        <strong
          class="MuiTypography-root sc-dlfnbm MuiTypography-subtitle2 MuiTypography-colorSecondary"
          data-testid="transaction_header_quantity"
        >
          123
        </strong>
      </div>
      <div>
        <p
          class="MuiTypography-root sc-gsTCUz MuiTypography-body2"
        >
          Valor total
        </p>
        <strong
          class="MuiTypography-root sc-dlfnbm MuiTypography-subtitle2 MuiTypography-colorSecondary"
          data-testid="transaction_header_total"
        >
          R$ 456,00
        </strong>
      </div>
    </div>
  </div>
</body>
`;
